Converted StringInquirer into a PORO with better performance.

taq opened the pull request:
Currently, `ActiveSupport::StringInquirer` inherit from `String`, and use `method_missing` on all of its calls, so we can have a performance problem.

Converting it to a `PORO` and creating a method on the instance with the name of the positive result and leaving just a test to return false to predicate methods can speed up operations.

The current benchmark result returned for the new benchmark test is:

```
all true  0.046796   0.000000   0.046796 (  0.047182)
all false 0.045121   0.000017   0.045138 (  0.045299)
half      0.042590   0.000000   0.042590 (  0.042956)
```

with this change it results in:

```
all true  0.013614   0.000000   0.013614 (  0.013640)
all false 0.029838   0.000008   0.029846 (  0.030124)
half      0.017558   0.000000   0.017558 (  0.017680)
```

taq committed:
Converted StringInquirer into a PORO with better performance.

Currently, ActiveSupport::StringInquirer inherit from String, and use
method_missing on all of its calls, so we can have a performance
problem.

Converting it to a PORO and creating a method on the instance with the
name of the positive result and leaving just a test to return false to
predicate methods can speed up operations.

The current benchmark result returned for the new benchmark test is:

all true  0.046796   0.000000   0.046796 (  0.047182)
all false 0.045121   0.000017   0.045138 (  0.045299)
half      0.042590   0.000000   0.042590 (  0.042956)

with this change it results in:

all true  0.013614   0.000000   0.013614 (  0.013640)
all false 0.029838   0.000008   0.029846 (  0.030124)
half      0.017558   0.000000   0.017558 (  0.017680)

casperisfine commented:
This isn't backward compatible. `StringInquirer` behaves like a `String` in all cases. Your change breaks that.

taq commented:
Thanks for your answer. Some questions:

- It really needs to inherit from String to do what it is supposed to do, check the string like a predicate?

- If so, should it have more tests to make sure all the String behaviors are there?

- If I change to inherit from String and keep the optimizations, is this an acceptable change?

Thanks.

casperisfine commented:
> * It really needs to inherit from String to do what it is supposed to do, check the string like a predicate?

It Inherit from String, so that it behaves like a String.

e.g. `Rails.env` is a `StringInquirer`, you can do `Rails.env.test?` but also `p Rails.env` or `Rails.env.upcase`.

In theory you could do this with composition, but it's tricky to do it in a purely backward compatible way, there's lots of footguns. IMO it's not worth it for something that's not meant to be in a hotspot.

> If so, should it have more tests to make sure all the String behaviors are there?

I mean, the whole CI is red, there's probably enough tests, even though some are indirect.

> If I change to inherit from String and keep the optimizations, is this an acceptable change?

No, because your optimization is entirely incorrect.

```ruby
    def initialize(method_name)
      @method_name = method_name

      self.class.send(:define_method, "#{method_name}?") { true }
    end
```

This mean every time you instantiate a new `StringInquirer`, you break all existing ones, e.g.:

```ruby
a = StringInquirer.new("admin")
a.admin? # => true
b = StringInquirer.new("user")
a.user? # => true
b.admin? # => true
```

So I don't think there's anything to keep in this PR.

taq commented:
Thanks for your answer.

> e.g. Rails.env is a StringInquirer, you can do Rails.env.test? but also p Rails.env or Rails.env.upcase.

Ops, I forgot that we can use it like this. Most of time I use the predicate form. Thanks for clarifying.

> In theory you could do this with composition, but it's tricky to do it in a purely backward compatible way, there's lots of footguns. IMO it's not worth it for something that's not meant to be in a hotspot.

Agree. 

> I mean, the whole CI is red, there's probably enough tests, even though some are indirect.

That's the point. Indirect tests most of time don't make explicit what they need. If we have some tests on `string_inquirer_test.rb` explicitly showing that it **needs** to be a `String`, I (and others) won't be asking if it needs to inherit from `String` and just running the specific test on this file will make sure all the rest is ok. Maybe just a `assert_kind_of` test there make things very clear.

> No, because your optimization is entirely incorrect.

Ouch, my bad. Didn't notice it was going to the class and not the instance, was in a hurry and need to pay more attention. Just changing there to 

```ruby
define_singleton_method("#{method_name}?") { true }
```

can fix this. Not entirely incorrect, I beg to differ. But the main idea there I'd want to be considered is creating an instance method to return right without calling `method_missing` to make performance better, I think is still a good idea.

So, I just changed the local code here to make it inherit from `String` again and fixed the instance method, and when compared to the current behavior it is still faster:

```
current behaviour
-----------------
all true  0.043308   0.000000   0.043308 (  0.043410)
all false 0.042461   0.000000   0.042461 (  0.042466)
half      0.040245   0.000000   0.040245 (  0.040250)
```

```
changed
-------
all true  0.009602   0.000000   0.009602 (  0.009742)
all false 0.026115   0.000162   0.026277 (  0.026443)
half      0.017090   0.000000   0.017090 (  0.017176)
```

Please note the least performance gain is right when `method_missing` is called all the times. 

If I make a new pull request with those changes and some more tests making it direct/explicity showing that it needs to inherit from a `String`, will you consider it?

Thanks again.

taq commented:
@matthewd not really on a hot path, just wondering about the cost of the method cache invalidation vs the `method_missing` call every time it needs to be checked.

matthewd commented:
> If `StringInquirer` was solely used for `Rails.env`

Notably, `Rails.env` isn't a `StringInquirer` at all: it's an [`EnvironmentInquirer`](https://github.com/rails/rails/blob/9527e5e3076b4b9b9418823540dc20336f095ae7/activesupport/lib/active_support/environment_inquirer.rb#L7), which already does most of this optimization.

taq commented:
> That's my main worry. If StringInquirer was solely used for Rails.env, eagerly defining a singleton method may make sense, but some people use it for ephemeral objects where this approach would absolutely trash performance.

Yeah, I just saw a lot of other uses also. But even on that case you think the singleton method is worst than calling `method_missing` every time, after the method lookup finishes? It's an interesting question ... 

> Overall StringInquirer isn't a big concern to me, so I'm not really keen on modifying it for performance unless it's made clear some people need it to perform better.

Ok. My original motivation was to try to make it better anyway. :-)

> Notably, Rails.env isn't a StringInquirer at all: it's an [EnvironmentInquirer](https://github.com/rails/rails/blob/9527e5e3076b4b9b9418823540dc20336f095ae7/activesupport/lib/active_support/environment_inquirer.rb#L7), which already does most of this optimization.

That inherits from `StringInquirer` and make the optimizations creating boolean instance variables that make the work of the instance method I proposed. Maybe using instance variables perform better than the instance methods. What a good discussion. :-)

casperisfine commented:
> Notably, `Rails.env` isn't a `StringInquirer` at all: it's an [`EnvironmentInquirer`](https://github.com/rails/rails/blob/9527e5e3076b4b9b9418823540dc20336f095ae7/activesupport/lib/active_support/environment_inquirer.rb#L7)

Ah right, I forgot about that one https://github.com/rails/rails/commit/09e037253238c61fdf587315ecdff8d56dca7d83

taq commented:
> Ah right, I forgot about that one https://github.com/rails/rails/commit/09e037253238c61fdf587315ecdff8d56dca7d83

See? Someone else was also worried about `method_missing`. And was Charles Nutter. :-D
